rosbag2_cpp/src/rosbag2_cpp/player_clock.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <mutex>
+#include <thread>
+
+#include "rosbag2_cpp/player_clock.hpp"
+#include "rosbag2_cpp/types.hpp"
+
+namespace
+{
+
+/**
+ * Stores an exact time match between a system steady clock and the semantic play clock.
+ * This is created whenever a factor changes such that a new base reference is needed
+ * such as pause, resume, rate change, or jump
+ */
+struct TimeSync
+{
+  rosbag2_cpp::PlayerClock::PlayerTimePoint player_time;
+  rosbag2_cpp::PlayerClock::RealTimePoint real_time;
+};
+
+template<typename T>
+rcutils_duration_value_t duration_nanos(const T & duration)
+{
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(duration).count();
+}
+
+}  // namespace
+
+namespace rosbag2_cpp
+{
+
+class PlayerClockImpl
```

Comment:
@emersonknapp It looks like `PlayerClockImpl` class is redundant (at least in current implementation) since it is pure POD type and unlikely to be replaced. \
I think it would  be much simpler to bring all incapsulated variables directly to the `PlayerClock` class and create them on the stack since they are small in size.

Reply:
It will be expanded very soon in followup PRs - I am inclined to present only a pure interface in the headers and hide the implementation via Pimpl pattern, because I think we have exposed far too much in headers in the past, making backports impossible.